### PR TITLE
More objectives

### PR DESCRIPTION
**Pull request by Botev**
So I wanted to add accuracy objective (so that it can be made more general of calculating validation and test statistics such as just passing a handle) as well as hinge loss objectives (I've seen them used often in the NLP papers). Additionally, changed the rectifier to use the theano implementation, because as of the current state they are the same.

### CODE REVIEW

File: lasagne/objectives.py [hunk 1]
```diff
@@ -198,3 +202,145 @@ def aggregate(loss, weights=None, mode='mean'):
...
+    -------
+    Theano 1D tensor
+        An expression for calculating the element-wise binary accuracy for each
+        data point
+
+    Notes
+    -----
+    This is a strictly non differential function as it includes an argmax.
+    This objective function should never be used with a gradient calculation.
+    It is intended as a convenience for validation and testing not training.
```
Comment (f0k) [inline lasagne/objectives.py]: Almost, there's no argmax in here :) For `categorical_accuracy`, the text is fine, but for `binary_accuracy`, what about:
```
This objective function should not be used with a gradient calculation;
its gradient is zero everywhere. It is intended as a convenience for
validation and testing, not training.
```

File: lasagne/objectives.py [hunk 2]
```diff
@@ -198,3 +202,145 @@ def aggregate(loss, weights=None, mode='mean'):
...
+    predictions : Theano 1D tensor
+        Predictions in [0, 1], such as a sigmoidal output of a neural network,
+        with data points in rows and a single column representing the
+        probability of the point belonging to the positive class
+    targets : Theano 1D tensor
+        Targets in {0, 1}, such as ground truth labels.
+    threshold : A scalar, default: 0.5
+        Specifies at what threshold to consider the predictions being of the
+        positive class
+    Returns
```
Comment (f0k) [inline lasagne/objectives.py]: missing newline before `Returns`

File: lasagne/objectives.py [hunk 3]
```diff
@@ -198,3 +202,145 @@ def aggregate(loss, weights=None, mode='mean'):
...
+    predictions : Theano 1D tensor
+        Predictions in (0, 1), such as sigmoidal output of a neural network.
+    targets : Theano tensor
+        Targets in {0, 1}, such as ground truth labels or {-1,1}, indicated
+        by the binary parameter.
+    binary : bool, default True
+        True if targets are in {0,1} False if they are in {-1,1}
+    delta : scalar, default 1
+        The hinge loss margin
+    Returns
```
Comment (f0k) [inline lasagne/objectives.py]: missing newline before `Returns`

File: lasagne/objectives.py [hunk 4]
```diff
@@ -198,3 +202,145 @@ def aggregate(loss, weights=None, mode='mean'):
...
+    predictions : Theano 2D tensor
+        Predictions in (0, 1), such as softmax output of a neural network,
+        with data points in rows and class probabilities in columns.
+    targets : Theano 2D tensor or 1D tensor
+        Either a vector of int giving the correct class index per data point
+        or a 2D tensor of 1 hot encoding of the correct class in the same
+        layout as predictions
+    delta : scalar, default 1
+        The hinge loss margin
+    Returns
```
Comment (f0k) [inline lasagne/objectives.py]: missing newline before `Returns`

**Comment by f0k** [comment]
Thank you, looks good except for minor comments (although I didn't review the tests yet).

Could you remove the `scaled_tanh` commit, though? We've already got a separate PR for that (I haven't forgotten, sorry for taking so long). The easiest way to remove it and squash the rest would be:
```
git reset --hard  7eed54b
git cherry-pick 45d1fe0
git cherry-pick ecc32bb
git reset --soft 45d1fe0
# now would be the time to make the edits I suggested,
# i.e., insert the missing empty lines and change the docstring
git add lasagne/objectives.py
git commit --amend -m "Added hinge loss and accuracy objective functions"
git push --force
```

Thank you!

File: lasagne/nonlinearities.py [hunk 5]
```diff
@@ -46,12 +46,10 @@ def softmax(x):
 # tanh
 def tanh(x):
     """Tanh activation function :math:`\\varphi(x) = \\tanh(x)`
-
```
Comment (benanne) [inline lasagne/nonlinearities.py]: why did you remove this whitespace? It's inconsistent with the other functions now.

**Comment by f0k** [comment]
Ah, you've already addressed my comments now, that was quick! The sequence should now be:
```
git reset --hard 7eed54b
git cherry-pick 45d1fe0
git cherry-pick ecc32bb
git cherry-pick bd7362e
git reset --soft 45d1fe0
git commit --amend -m "Added hinge loss and accuracy objective functions"
git push --force
```

File: lasagne/nonlinearities.py [hunk 6]
```diff
@@ -78,7 +142,7 @@ def rectify(x):
     # and it works with nonsymbolic inputs as well.
     # Thanks to @SnipyHollow for pointing this out. Also see:
     # https://github.com/Lasagne/Lasagne/pull/163#issuecomment-81765117
-    return 0.5 * (x + abs(x))
+    return theano.tensor.nnet.relu(x)
```
Comment (benanne) [inline lasagne/nonlinearities.py]: perhaps the comments above should be removed as well, since they refer to the particular implementation we were using.

File: lasagne/nonlinearities.py (hunk 5 above)
Comment (f0k) [inline lasagne/nonlinearities.py]: Don't worry, that commit will be removed, it's part of a different PR (in which it also hasn't been addressed yet: https://github.com/Lasagne/Lasagne/pull/409#discussion_r39526174, but I guess I'll take care of it myself).

**Comment by Botev** [comment]
Sorry, that commit shouldn't have been here. Cause I branched from the master of my fork... The tests are relatively simple, but take a look before merging.

**Comment by f0k** [comment]
> Sorry, that commit shouldn't have been here.

But now it's still there... it seems you didn't follow the steps I posted, but added another set of changes that undid 87f582e and merged it with the others. Can you please just follow the steps from https://github.com/Lasagne/Lasagne/pull/428#issuecomment-142254790? Ideally, everything considering the rectifier would be a separate commit, but by now I'm confused what was in which commit...

**Comment by Botev** [comment]
I followed exactly the ones posted (as from my history):
```
 1999  git reset --hard 7eed54b
 2000  git cherry-pick 45d1fe0
 2001  git cherry-pick ecc32bb
 2002  git cherry-pick bd7362e
 2003  git reset --soft 45d1fe0
 2004  git commit --amend -m "Added hinge loss and accuracy objective functions"
 2005  git push --force
```
I thought that was the intend, to keep the ScaledTanh in the other branch until its ready for the merge (I think only the comment need fixing). As whether the relu is a separate commit not sure...

**Comment by f0k** [comment]
> I followed exactly the ones posted [...]
> I thought that was the intend, to keep the ScaledTanh in the other branch until its ready for the merge

Yes, exactly. But 87f582e still shows up in this PR (#428) as well. Aaah, wait, now I see why... sorry, my mistake! It should be:
```
git reset --hard 7eed54b
git cherry-pick 45d1fe0
git cherry-pick ecc32bb
git cherry-pick bd7362e
git reset --soft HEAD~2  # resetting to 45d1fe0 was wrong, that's still based on the old history
git commit --amend -m "Added hinge loss and accuracy objective functions"
git push --force
```

> As whether the relu (removing the comments) is 1 commit after that.

For a really clean history, there should be one commit adding your objectives along with tests (that's one chunk of work), and another commit replacing relu with Theano's implementation and removing the old relu comments (that's another chunk of work). If you're fine with sorting them out, let me know and I'll guide you. Otherwise I'd be fine with merging everything into a single commit. That would be (after the above):
```
git cherry-pick c383c51
git reset --soft HEAD^
git commit --amend -no-edit
git push --force
```

**Comment by Botev** [comment]
I'm fine with getting more lessons on git so if you know how I'm happy to sort out the history. I mean tbh at this point I can just copy paste the stuff and fork the main branch (seems easier), but if you know a better way I'm happy to try.

**Comment by Botev** [comment]
Well I tried something, but it did not work as expected apparently :/

**Comment by f0k** [comment]
> Well I tried something, but it did not work as expected apparently :/

Don't worry! At least we've now got all of the objective stuff in one commit, except for the "Fixed missing new lines."
So the first thing to learn would be: When you do a minor change to an existing set of work, it's better to amend the commit you had. The general way would be:
```
# first, edit the files
git add ...  # then, stage the changes
git commit --amend --no-edit  # now, instead of adding a new commit, amend the existing one
git push --force  # whenever you amend something, you need a force push
```

To merge the last 2 commits into one, do the following:
```
git reset --soft HEAD~1  # go one step back in history, but keep all changes staged
git commit --amend  # this merges the changes into the commit you're pointing at
```

About the relu changes: I think it will be the easiest to redo them now, and then commit them (this time as a separate commit, because that change doesn't have anything to do with the objective functions). I don't know which commit they were originally included in, otherwise you could do `git cherry-pick ...`.

Why I'm so keen on a clean history is that it makes it a lot easier to track down breaking changes later -- even with 100% test coverage, it can happen that some change introduces some bug, and it's a lot easier if changes are grouped into consistent packages (commits) than if they're mixing different things or only doing half of the work. Thanks for bearing with me :)

**Comment by Botev** [comment]
Ok so I think now it should be done.

**Comment by f0k** [comment]
Perfect, thanks a lot! Merging.

**Comment by benanne** [comment]
Nice, thanks for contributing!
